extensions/emmet/src/abbreviationActions.ts

```diff
@@ -216,66 +202,71 @@ function doWrapping(_: boolean, args: any) {
 		}, { undoStopBefore: false, undoStopAfter: false });
 	}
 
-	function makeChanges(inputAbbreviation: string | undefined, definitive: boolean): Thenable<boolean> {
+	let inPreviewMode = false;
+	async function makeChanges(inputAbbreviation: string | undefined, previewChanges: boolean): Promise<boolean> {
 		if (!inputAbbreviation || !inputAbbreviation.trim() || !helper.isAbbreviationValid(syntax, inputAbbreviation)) {
-			return inPreview ? revertPreview().then(() => { return false; }) : Promise.resolve(inPreview);
+			if (inPreviewMode) {
+				inPreviewMode = false;
+				await revertPreview();
+			}
+			return false;
 		}
 
 		const extractedResults = helper.extractAbbreviationFromText(inputAbbreviation);
 		if (!extractedResults) {
-			return Promise.resolve(inPreview);
-		} else if (extractedResults.abbreviation !== inputAbbreviation) {
-			// Not clear what should we do in this case. Warn the user? How?
+			if (inPreviewMode) {
+				inPreviewMode = false;
+				await revertPreview();
+			}
+			return false;
```

Comment:
For one of my previous comments, I meant that it seems this part repeats lines 208 to 212. I can see combining all the cases making it even harder to read, though, so I'm not sure whether it's worth changing this part.

Reply:
Pushed another commit merging the first two `if (inPreviewMode)` blocks